broadinstitute/scp-ingest-pipeline
# Minify ontologies and keep them fresh, for fast metadata validation (SCP-5719)

This lays a foundation for drastically faster ontology label validation.  Once integrated, that means easier upload for users.

### Background
Adding metadata for cells currently requires authors to specify ontology IDs and formal labels for certain annotations.  This enables faceted search across studies, consistent fields for exploration within studies, and potential machine-learning applications.  

Ensuring those IDs and labels are correct requires validation when data is added.  Failing this validation is common in the author journey, and at 15-45 minutes per attempt it's a frustratingly slow experience.  Instant validation would help!

### Technical overview
#### Minify ontologies
Previously, such ontology metadata was validated _after_ upload by querying EBI's Ontology Lookup Service (OLS).  Each unique ID-label pair was queried via a remote OLS API endpoint.  Results were cached per ingest run, but needing to upload data first and then querying a remote API wasn't especially fast.

Now, the ontologies used by OLS are available as small static files.  A new pipeline module fetches the latest versions of each ontology from its original source on GitHub.  The original ontology is then slimmed into a minimal TSV and compressed.  This converts ~224 MB in ontology JSON files into 2 MB TSV.GZs containing only the needed data: one row for each class, listing its ID, label, and synonyms.  

The smaller reference data will be quickly loadable into the client.  That unlocks doing this validation _before_ upload, and using in-memory data rather than remote queries.  The validation will speed up immensely -- 15-45 _minutes_ before, 10-15 _seconds_ after.

#### Keep them fresh
The minified ontologies are automatically updated.  When a PR is opened, the new "Minify ontologies" GitHub Action workflow runs `minify_ontologies.py` to fetch the newest versions of each ontology.  Note the new GitHub check at bottom of this PR.  The GHA workflow compares the newest version to the version in the repo, and [commits](https://github.com/broadinstitute/scp-ingest-pipeline/pull/364/commits/978d6f0d95d79a44eb641dd2a21a01a6b3bb3124) the new version if it differs.

### Next steps
These static files (e.g. `mondo.min.tsv.gz`) will be useful when they're integrated into client-side file validation (CSFV):

1.  On page load, fetch these ontologies, load into client memory.  They're a 2 MB transfer in total, so fast to load and small even decompressed and parsed.
2.  Upon selecting AnnData, apply [row-level validation](https://github.com/broadinstitute/single_cell_portal_core/pull/2122) for required metadata validation, using these small static ontologies for instant ID-label mapping

That general approach, sans fetch, would also work for ingest's backend Python.

### Test
Automated tests verify the basic processing flow.  To optionally manually test the GitHub Action:
1.  Get this branch:  `git fetch -a; git switch ew-minify-ontologies`
2.  Create a test branch: `git switch -c <your initials>-minify-ontologies-test`
3. `cd ingest/validation`
4.  Decompress mini MONDO: `gzip -dkf mondo.min.tsv.gz`
5.  Add some content to `mondo.min.tsv`
6.  Compress it: `gzip -kf mondo.min.tsv`
7. Open `.github/workflows/minify_ontologies.yml`
8. Uncomment the `push` trigger block, replace `ew-minify-ontologies` with your branch name
9. Save, commit, push
10.  Navigate to [Actions](https://github.com/broadinstitute/scp-ingest-pipeline/actions)
11.  Click into your "Minify ontology" workflow run, click "build", expand "Diff and commit changes"
12.  Confirm you see a message at the bottom broadly like the following:
```
[ew-minify-ontologies 978d6f0] Update minified ontologies via GitHub Actions
 2 files changed, 0 insertions(+), 0 deletions(-)
+ git push origin HEAD
To https://github.com/broadinstitute/scp-ingest-pipeline
   a62f7ea..978d6f0  HEAD -> ew-minify-ontologies
```
13.  View your branch's commits in GitHub
14. Confirm you see a commit from "github-actions" with a message like "Update minified ontologies via GitHub Actions"
15. Delete your branch

This relates to SCP-5719.
